### DISCUSSION TITLE
need IndexFlatIP support float16

### DISCUSSION THREAD

**Discussion by yangsp5** [facebookresearch/faiss#3539] [Q&A]
# Summary

need `IndexFlatIP` support `float16`

when the number of vector is very very large, such as `1e10`. Use `IndexFlatIP` of `float32` is too expensive, maybe `float16` is much fastter

**Comment by mdouze** [comment]
You can use `IndexFlatScalarQuantizer` with `QT_fp16`.

**Comment by yangsp5** [comment]
How to use `IndexFlatScalarQuantizer` with `QT_fp16`,

`index = faiss.IndexFlatScalarQuantizer(emb_size, faiss.ScalarQuantizer.QT_fp16)` got wrong

**Reply by dxsanny99** [reply]
facebook.com/profile.php?id=100054340704401

**Reply by dxsanny99** [reply]
facebook.com/profile.php?id=100054340704401

**Comment by yunjiangster** [comment]
@yangsp5 I think @mdouze meant `IndexScalarQuantizer`, but that still takes in float32 index vectors. I found the function `knn_gpu` that computes k nearest neighbors without building an index first, and it seems to support float16 input directly.

**Comment by ucas010** [comment]
**IndexScalarQuantizer** support  float16?
and P or R @1 is what ?
I also have 1e10 datasets

**Comment by dxsanny99** [comment]
acebook.com/profile.php?id=100054340704401

**Comment by mdouze** [comment]
The conversion from fp32 to fp16 is done internally. 
Note that most CPUs don't handle fp16 computations natively anyways so all computations are performed in float32.
knn_gpu supports float16 because there is hardware support for it.
